### PR TITLE
improvement: Realtime Markdown rendering

### PR DESCRIPTION
**Pull request by wasimsandhu**
## 📝 Summary

Completes #2137.

## 🔍 Description of Changes

realtime markdown

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [ ] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

File: frontend/src/core/codemirror/cm.ts [hunk 1]
```diff
@@ -87,6 +88,8 @@ export const setupCodeMirror = (opts: CodeMirrorSetupOpts): Extension[] => {
     // Cell editing
     cellMovementBundle(cellId, cellMovementCallbacks, hotkeys),
     cellCodeEditingBundle(cellId, cellCodeCallbacks, hotkeys),
+    // Markdown autorun
+    markdownAutoRunExtension(cellMovementCallbacks),
```
Comment (mscolnick) [inline frontend/src/core/codemirror/cm.ts:92]: Would this be better in the MarkdownLanguageAdapter class? Better separation of concerns and no need to do to the if check

**Comment by dmadisetti** [comment]
Nice! Might be worth seeing how this interacts with the VSCode plugin which already does this

**Comment by mscolnick** [comment]
@dmadisetti - its essentially the exact same logic, but they won't overlap or effect each other.

**Comment by akshayka** [comment]
@wasimsandhu , this is awesome!!

**Review by akshayka — COMMENTED** [review]
I don't remember if we "debounce"/deduplicate run requests in the backend. I think we might not, in which case if the kernel were busy while you were typing in a markdown cell, it's possible that hundreds of run requests would get queued and then run serially.

Might be something to test -- do time.sleep(10) in one cell, then start typing a bunch of text in a markdown cell and see what happens. I can help improve the BE if this is indeed an issue.

**Comment by wasimsandhu** [comment]
Ah that's a good point. Let me check it out and report back

**Comment by mscolnick** [comment]
@wasimsandhu - its ok if the markdown is queued and doesn't update. but the worry is if after the sleep is done, will it run 30 requests (one for each char) through the kernel or 1. ideally its 1, if that is not the case, I will dedupe RunRequests in another PR.

**Comment by wasimsandhu** [comment]
@mscolnick Ahhh gotcha. If it's alright, I'd love to take a stab at it

**Comment by mscolnick** [comment]
Totally - go for it! Thanks

**Comment by wasimsandhu** [comment]
Just finished implementing @akshayka's de-duplication logic only to realize run requests were not getting duplicated to begin with... the markdown cell is stale while other cells are running, and it does not autorun after the `time.sleep` cell finishes executing.

I guess I must have hallucinated it the first time I observed it? I was positive I saw tons of duplicated requests, but now I'm unable to reproduce it lol. Can someone test on their machine and confirm that requests are not being duplicated as the branch currently stands?

**Comment by wasimsandhu** [comment]
EDIT: Please ignore my previous message... I just restarted the server and reproduced the request duplication. Feels like I just got gaslighted

**Comment by wasimsandhu** [comment]
Looks like I misled myself. After trying some different things, I realized that:

1) multiple run requests are not being sent to the server after blocking cell runs
2) run requests were being sent to the server on *every* interaction with the cell (click, drag, etc.)

Updated extensions.ts so that a run only occurs if the cell's code was changed 😅

**Comment by akshayka** [comment]
Sounds good. I will review tomorrow

**Comment by mscolnick** [comment]
@wasimsandhu i just tested it. you are correct, it works as expected and we aren't sending additional requests

**Review by mscolnick — APPROVED** [review]
amazing! such a nice change

**Pull request merged**